File: 42-TrappingRainWater/42-TrappingRainWater.cpp
// Last updated: 6/11/2026, 11:37:56 AM
class Solution {
public:
    int trap(vector<int>& arr) {
        int n = arr.size();
        vector<int> nge(n, -1);
        vector<int> pge(n, -1);
        nge[n-1] = n-1;
        for(int i=n-2; i>=0; i--){
            if(arr[nge[i+1]]>arr[i]){
                nge[i] = nge[i+1];
            }else{
                nge[i] = i;
            }
        }

        pge[0] = 0;
        for(int i=1; i<n; i++){
            if(arr[pge[i-1]] > arr[i]){
                pge[i] = pge[i-1];
            }else{
                pge[i] = i;
            }
        }

        int ans = 0;
        for(int i=1; i<n-1; i++){
            int pg = pge[i];
            int ng = nge[i];
            if(pg == i || i == ng) continue;
            int area = min(arr[pg], arr[ng]) - arr[i];
            ans += (area);
        }

        return ans;
    }
};

File: 42-TrappingRainWater/42-TrappingRainWater.py
# Last updated: 6/11/2026, 11:38:01 AM
class Solution:
    def trap(self, height: List[int]) -> int:
        twt = 0

        lm = [0]*(len(height))
        rm = [0]*(len(height))

        lm[0] = 0
        for i in range(1,len(height)):
            lm[i] = max(lm[i-1], height[i-1])

        rm[len(height) -1] = 0
        for i in range(len(height)-2, -1, -1):
            rm[i] = max(rm[i+1], height[i+1])


        for i in range(1,len(height)-1):
            curh = height[i]
            # lefth = 0
            # for j in range(0, i):
            #     lefth = max(lefth, height[j])
            # righth = 0
            # for j in range(len(height)-1, i, -1):
            #     righth = max(righth, height[j])

            wt = (min(lm[i], rm[i]) - curh)
            if(wt>0):
                twt += wt
        return twt